app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/AccountController.scala

```diff
@@ -48,69 +52,61 @@ class AccountController @Inject() (
+      case _ => Redirect(routes.UpdateEmailAddressController.updateEmailAddress()).toFuture
     }
-    retrieveEmailService.retrieveEmailByEORI(eori).flatMap {
-      _ match {
-        case Some(_) => getAccountFlow
-        case None => Future.successful(Redirect(routes.UpdateEmailAddressController.updateEmailAddress()))
-      }
+  }
+
+  private def getUndertakingAndHandleResponse(implicit r: EscAuthRequest[AnyContent], e: EORI): Future[Result] = {
+    escService.retrieveUndertaking(r.eoriNumber) flatMap {
+      case Some(u) => handleExistingUndertaking(u)
+      case None => handleUndertakingNotCreated
     }
   }
 
-  def getExistingUndertaking: Action[AnyContent] = escAuthentication.async { implicit request =>
-    implicit val eori: EORI = request.eoriNumber
+  private def handleUndertakingNotCreated(implicit e: EORI): Future[Result] = {
+    val result = getOrCreateJourneys().map {
+      case (ej, uj) if !ej.isComplete && uj.isEmpty => Redirect(routes.EligibilityController.firstEmptyPage())
+      case (_, uj) if !uj.isComplete => Redirect(routes.UndertakingController.firstEmptyPage())
+      case _ => Redirect(routes.BusinessEntityController.getAddBusinessEntity())
+    }
+    result.getOrElse(handleMissingSessionData("Account Home - Undertaking not created -"))
+  }
+
+  private def handleExistingUndertaking(u: Undertaking)(implicit r: EscAuthRequest[AnyContent], e: EORI): Future[Result] = {
+    val result = for {
+      _ <- store.put(u).toContext
+      _ <- getOrCreateJourneys(UndertakingJourney.fromUndertaking(u))
+    } yield renderAccountPage(u)
+
+    result.getOrElse(handleMissingSessionData("Account Home - Existing Undertaking -"))
+  }
+
+  private def getOrCreateJourneys(u: UndertakingJourney = UndertakingJourney())(implicit  e: EORI) =
     for {
-      undertakingOpt <- escService.retrieveUndertaking(eori)
-    } yield undertakingOpt match {
-      case Some(undertaking) if undertaking.isLeadEORI(eori) =>
-        Redirect(routes.AccountController.getAccountPage()) //if logged in as lead EORI, redirect to Account home page
-      case Some(undertaking) =>
-        Ok(
-          existingUndertakingPage(undertaking.name)
-        ) //if logged in as non-lead EORI, redirect to existing undertaking page
-      case None =>
-        Redirect(routes.AccountController.getAccountPage()) //if undertaking not present, then create undertaking
+      ej <- store.get[EligibilityJourney].toContext.orElse(store.put(EligibilityJourney()).toContext)
+      uj <- store.get[UndertakingJourney].toContext.orElse(store.put(u).toContext)
+      _ <- store.get[BusinessEntityJourney].toContext.orElse(store.put(BusinessEntityJourney()).toContext)
+    } yield (ej, uj)
+
+  private def renderAccountPage(undertaking: Undertaking)(implicit r: EscAuthRequest[AnyContent]) = {
+    val currentTime = timeProvider.today
+
+    val isTimeToReport = ReportReminderHelpers.isTimeToReport(undertaking.lastSubsidyUsageUpdt, currentTime)
```

Comment:
We don't need to calculate all this stuff if it's a non lead EORI, can we do all this calculation under the if clause?

Reply:
There's another ticket coming where we will need to pass dates to the non-lead page too so I figured it was best to leave as is to make that future work a little easier.